### PR TITLE
add mw80 keyboard

### PR DESCRIPTION
**Pull request by TW59420**
## Description
MWStudio is a custom keyboard enthusiast
Upload qmk firmware of mw80 PCB

## Types of Changes
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist
- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/mwstudio/mw80/apm32f103/config.h [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+#define DIODE_DIRECTION COL2ROW
+
+/* Set 0 if debouncing isn't needed */
+#define DEBOUNCE 5
+
+#define FORCE_NKRO
+
+/* 将USB 轮询速率更改为 1000hz 并为精英游戏每次扫描使用更大的密钥*/
+#define USB_POLLING_INTERVAL_MS  1
+#define QMK_KEYS_PER_SCAN  12
```
Comment (drashna) [inline keyboards/mwstudio/mw80/apm32f103/config.h:49]: Not necessary.
```suggestion
```

File: keyboards/mwstudio/mw80/apm32f103/rules.mk [hunk 2]
```diff
@@ -0,0 +1,10 @@
+# MCU name
+MCU = STM32F103
+
+# Bootloader selection
+# BOOTLOADER = tinyuf2
```
Comment (drashna) [inline keyboards/mwstudio/mw80/apm32f103/rules.mk:5]:
```suggestion
```

File: keyboards/mwstudio/mw80/keymaps/default/keymap.c [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    ),
+
+	[3] = LAYOUT(
+		KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
+		KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
+		KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
+		KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,          KC_TRNS,
+		KC_TRNS,          KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,          KC_TRNS,          KC_TRNS,
+		KC_TRNS, KC_TRNS, KC_TRNS,                            KC_TRNS,                            KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS
+    ),
```
Comment (drashna) [inline keyboards/mwstudio/mw80/keymaps/default/keymap.c:54]: These are completely empty, correct?  If so, they're not really needed on a non-via keymap.
```suggestion
```

File: keyboards/mwstudio/mw80/keymaps/via/rules.mk [hunk 4]
```diff
@@ -0,0 +1,2 @@
+VIA_ENABLE = yes
+LTO_ENABLE = yes
```
Comment (drashna) [inline keyboards/mwstudio/mw80/keymaps/via/rules.mk:2]: Not really needed, as space isn't a limiting factor here?
```suggestion
```

File: keyboards/mwstudio/mw80/readme.md [hunk 5]
```diff
@@ -0,0 +1,22 @@
+# MW80
+
+![MW80](https://i.imgur.com/ygcLrhz.png)
```
Comment (drashna) [inline keyboards/mwstudio/mw80/readme.md:3]:
```suggestion
![MW80](https://i.imgur.com/ygcLrhzh.png)
```

File: keyboards/mwstudio/mw80/rules.mk [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+#
+BOOTMAGIC_ENABLE = yes      # Enable Bootmagic Lite
+MOUSEKEY_ENABLE = yes       # Mouse keys
+EXTRAKEY_ENABLE = yes       # Audio control and System control
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = no         # Commands for debug and configuration
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no      # Breathing sleep LED during USB suspend
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = yes           # USB Nkey Rollover
```
Comment (drashna) [inline keyboards/mwstudio/mw80/rules.mk:12]: Please remove the `SLEEP_LED_ENABLE` setting, as it's not really used, and remove the comments for both it and NKRO.  
```suggestion
NKRO_ENABLE = no            # Enable N-Key Rollover
```

File: keyboards/mwstudio/mw80/mw80.c [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    { 224, 32 }, { 179, 32 }, { 134, 32 }, { 90, 32 }, { 45, 32 }, { 0, 32 },
+    { 224, 32 }, { 199, 32 }, { 174, 32 }, { 149, 32 }, { 125, 32 }, { 100, 32 }, { 75, 32 }, { 50, 32 }, { 25, 32 }, { 0, 32 }
+    //这一部分对应灯珠的物理矩阵位置，按顺序写
+}, {
+    1, 1, 1, 1, 1, 1,
+    3, 3, 3, 3, 3, 3, 3, 3, 3, 3
+    //标注灯珠类型，一般把轴灯和底灯分开即可
+} };
+#endif
+
```
Comment (drashna) [inline keyboards/mwstudio/mw80/mw80.c:32]: This allows for a default behavior to be defined, without the need for the user to specify the behavior.  This is very useful for the configurator which doesn't have a way to configure this behavior, currently.

```suggestion

#ifdef ENCODER_ENABLE
bool encoder_update_kb(uint8_t index, bool clockwise) {
    if (!encoder_update_user(index, clockwise)) { return false; }
    if (index == 0) {
        if (clockwise) {
            tap_code_delay(KC_VOLU, 10);
        } else {
            tap_code_delay(KC_VOLD, 10);
        }
    }
    return true;
}
#endif
```

**Review by drashna — CHANGES_REQUESTED** [review]

**Comment by TW59420** [comment]
Thank you for your help

**Review by drashna — APPROVED** [review]

File: keyboards/mwstudio/mw80/mw80.c [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+led_config_t g_led_config = { {
+
+}, {
+    { 224, 32 }, { 179, 32 }, { 134, 32 }, { 90, 32 }, { 45, 32 }, { 0, 32 },
+    { 224, 32 }, { 199, 32 }, { 174, 32 }, { 149, 32 }, { 125, 32 }, { 100, 32 }, { 75, 32 }, { 50, 32 }, { 25, 32 }, { 0, 32 }
+    //这一部分对应灯珠的物理矩阵位置，按顺序写
+}, {
+    1, 1, 1, 1, 1, 1,
+    3, 3, 3, 3, 3, 3, 3, 3, 3, 3
+    //标注灯珠类型，一般把轴灯和底灯分开即可
```
Comment (fauxpark) [inline keyboards/mwstudio/mw80/mw80.c:29]:
```suggestion
}, {
    1, 1, 1, 1, 1, 1,
    3, 3, 3, 3, 3, 3, 3, 3, 3, 3
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/mwstudio/mw80/mw80.c [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "mw80.h"
+
+#ifdef RGB_MATRIX_ENABLE
+led_config_t g_led_config = { {
+
```
Comment (zvecr) [inline keyboards/mwstudio/mw80/mw80.c:21]: This section of the LED config should not be blank, and should match matrix width/height.

**Review by zvecr — CHANGES_REQUESTED** [review]
Also, missing a `info.json` <https://docs.qmk.fm/#/reference_info_json>

File: keyboards/mwstudio/mw80/mw80.c (hunk 9 above)
Comment (TW59420) [reply keyboards/mwstudio/mw80/mw80.c:21]: no led in matrix, only mood lights

File: keyboards/mwstudio/mw80/info.json [hunk 10]
```diff
@@ -0,0 +1,104 @@
+{
+    "keyboard_name": "MW80",
+    "url": "",
+    "maintainer": "qmk",
```
Comment (fauxpark) [inline keyboards/mwstudio/mw80/info.json:4]:
```suggestion
    "keyboard_name": "MW80",
    "manufacturer": "MWStudio",
    "url": "",
    "maintainer": "qmk",
    "usb": {
        "vid": "0x7BA1",
        "pid": "0x8001",
        "device_version": "0.0.1"
    },
```

File: keyboards/mwstudio/mw80/apm32f103/config.h [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0x7BA1
+#define PRODUCT_ID      0x8001
+#define DEVICE_VER      0x0001
+#define MANUFACTURER    MWStudio
+#define PRODUCT         MW80
+
```
Comment (fauxpark) [inline keyboards/mwstudio/mw80/apm32f103/config.h:27]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]
Please ensure all files are using Unix line endings.

**Review by fauxpark — CHANGES_REQUESTED** [review]
`info.json` still has Windows line endings.

**Review by fauxpark — CHANGES_REQUESTED** [review]
.

**Review by drashna — APPROVED** [review]

File: keyboards/mwstudio/mw80/info.json [hunk 12]
```diff
@@ -0,0 +1,119 @@
+{
```
Comment (fauxpark) [inline keyboards/mwstudio/mw80/info.json:1]: This file needs to use Unix line endings (LF), not windows (CRLF). You must do this locally; the GitHub web editor cannot change it.
image

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/mwstudio/mw80/mw80.c (hunk 9 above)
Comment (fauxpark) [reply keyboards/mwstudio/mw80/mw80.c:21]: It still can't be blank.

File: keyboards/mwstudio/mw80/keymaps/default/keymap.c [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+		KC_ESC, KC_F1, KC_F2, KC_F3, KC_F4, KC_F5, KC_F6, KC_F7, KC_F8,   KC_F9, KC_F10,  KC_F11,  KC_F12,  KC_F13,  KC_PSCR, KC_SLCK, KC_PAUS,
+		KC_GRV, KC_1,  KC_2,  KC_3,  KC_4,  KC_5,  KC_6,  KC_7,  KC_8,    KC_9,  KC_0,    KC_MINS, KC_EQL,  KC_BSPC, KC_INS,  KC_HOME, KC_PGUP,
+		KC_TAB, KC_Q,  KC_W,  KC_E,  KC_R,  KC_T,  KC_Y,  KC_U,  KC_I,    KC_O,  KC_P,    KC_LBRC, KC_RBRC, KC_BSLS, KC_DEL,  KC_END,  KC_PGDN,
+		KC_CAPS,KC_A,  KC_S,  KC_D,  KC_F,  KC_G,  KC_H,  KC_J,  KC_K,    KC_L,  KC_SCLN, KC_QUOT,          KC_ENT,
+		KC_LSFT,       KC_Z,  KC_X,  KC_C,  KC_V,  KC_B,  KC_N,  KC_M,  KC_COMM, KC_DOT,  KC_SLSH,          KC_LSFT,          KC_UP,
+		KC_LCTL,KC_LGUI,KC_LALT,                   KC_SPC,                       KC_LALT, KC_LGUI, MO(1),   KC_LCTL, KC_LEFT, KC_DOWN, KC_RIGHT
+    ),
+
+	[1] = LAYOUT(
+		RESET,   KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
```
Comment (fauxpark) [inline keyboards/mwstudio/mw80/keymaps/default/keymap.c:30]:
```suggestion
		QK_BOOT,  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
```

File: keyboards/mwstudio/mw80/keymaps/default/keymap.c [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include QMK_KEYBOARD_H
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+
+	[0] = LAYOUT(
+		KC_ESC, KC_F1, KC_F2, KC_F3, KC_F4, KC_F5, KC_F6, KC_F7, KC_F8,   KC_F9, KC_F10,  KC_F11,  KC_F12,  KC_F13,  KC_PSCR, KC_SLCK, KC_PAUS,
```
Comment (fauxpark) [inline keyboards/mwstudio/mw80/keymaps/default/keymap.c:21]:
```suggestion
		KC_ESC, KC_F1, KC_F2, KC_F3, KC_F4, KC_F5, KC_F6, KC_F7, KC_F8,   KC_F9, KC_F10,  KC_F11,  KC_F12,  KC_F13,  KC_PSCR, KC_SCR, KC_PAUS,
```

File: keyboards/mwstudio/mw80/keymaps/default/keymap.c [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+		KC_GRV, KC_1,  KC_2,  KC_3,  KC_4,  KC_5,  KC_6,  KC_7,  KC_8,    KC_9,  KC_0,    KC_MINS, KC_EQL,  KC_BSPC, KC_INS,  KC_HOME, KC_PGUP,
+		KC_TAB, KC_Q,  KC_W,  KC_E,  KC_R,  KC_T,  KC_Y,  KC_U,  KC_I,    KC_O,  KC_P,    KC_LBRC, KC_RBRC, KC_BSLS, KC_DEL,  KC_END,  KC_PGDN,
+		KC_CAPS,KC_A,  KC_S,  KC_D,  KC_F,  KC_G,  KC_H,  KC_J,  KC_K,    KC_L,  KC_SCLN, KC_QUOT,          KC_ENT,
+		KC_LSFT,       KC_Z,  KC_X,  KC_C,  KC_V,  KC_B,  KC_N,  KC_M,  KC_COMM, KC_DOT,  KC_SLSH,          KC_LSFT,          KC_UP,
+		KC_LCTL,KC_LGUI,KC_LALT,                   KC_SPC,                       KC_LALT, KC_LGUI, MO(1),   KC_LCTL, KC_LEFT, KC_DOWN, KC_RIGHT
```
Comment (fauxpark) [inline keyboards/mwstudio/mw80/keymaps/default/keymap.c:26]:
```suggestion
		KC_LCTL,KC_LGUI,KC_LALT,                   KC_SPC,                       KC_LALT, KC_LGUI, MO(1),   KC_LCTL, KC_LEFT, KC_DOWN, KC_RGHT
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by fauxpark** [comment]
info.json is now missing...

File: keyboards/mwstudio/mw80/keymaps/via/keymap.c [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include QMK_KEYBOARD_H
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+
+	[0] = LAYOUT(
+		KC_ESC, KC_F1,  KC_F2, KC_F3, KC_F4, KC_F5, KC_F6, KC_F7, KC_F8,   KC_F9, KC_F10,  KC_F11,  KC_F12,  KC_F13,  KC_PSCR, KC_SLCK, KC_PAUS,
```
Comment (fauxpark) [inline keyboards/mwstudio/mw80/keymaps/via/keymap.c:22]:
```suggestion
		KC_ESC, KC_F1,  KC_F2, KC_F3, KC_F4, KC_F5, KC_F6, KC_F7, KC_F8,   KC_F9, KC_F10,  KC_F11,  KC_F12,  KC_F13,  KC_PSCR, KC_SCRL, KC_PAUS,
```

File: keyboards/mwstudio/mw80/keymaps/via/keymap.c [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+		KC_GRV, KC_1,   KC_2,  KC_3,  KC_4,  KC_5,  KC_6,  KC_7,  KC_8,    KC_9,  KC_0,    KC_MINS, KC_EQL,  KC_BSPC, KC_INS,  KC_HOME, KC_PGUP,
+		KC_TAB, KC_Q,   KC_W,  KC_E,  KC_R,  KC_T,  KC_Y,  KC_U,  KC_I,    KC_O,  KC_P,    KC_LBRC, KC_RBRC, KC_BSLS, KC_DEL,  KC_END,  KC_PGDN,
+		KC_CAPS,KC_A,   KC_S,  KC_D,  KC_F,  KC_G,  KC_H,  KC_J,  KC_K,    KC_L,  KC_SCLN, KC_QUOT,          KC_ENT,
+		KC_LSFT,        KC_Z,  KC_X,  KC_C,  KC_V,  KC_B,  KC_N,  KC_M,  KC_COMM, KC_DOT,  KC_SLSH,          KC_LSFT,          KC_UP,
+		KC_LCTL,KC_LGUI,KC_LALT,                   KC_SPC,                       KC_LALT, KC_LGUI, MO(1),   KC_LCTL, KC_LEFT, KC_DOWN, KC_RIGHT
```
Comment (fauxpark) [inline keyboards/mwstudio/mw80/keymaps/via/keymap.c:27]:
```suggestion
		KC_LCTL,KC_LGUI,KC_LALT,                   KC_SPC,                       KC_LALT, KC_LGUI, MO(1),   KC_LCTL, KC_LEFT, KC_DOWN, KC_RGHT
```

File: keyboards/mwstudio/mw80/keymaps/via/keymap.c [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+    ),
+
+	[1] = LAYOUT(
+		RESET,   KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
```
Comment (fauxpark) [inline keyboards/mwstudio/mw80/keymaps/via/keymap.c:31]:
```suggestion
		QK_BOOT,  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by TW59420** [comment]
re-uploaded info.json, use Unix line endings

**Comment by TW59420** [comment]
you change key code, qmk ci build is failing. Is that correct?

File: keyboards/mwstudio/mw80/keymaps/default/keymap.c [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include QMK_KEYBOARD_H
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+
+	[0] = LAYOUT(
+		KC_ESC, KC_F1, KC_F2, KC_F3, KC_F4, KC_F5, KC_F6, KC_F7, KC_F8,   KC_F9, KC_F10,  KC_F11,  KC_F12,  KC_F13,  KC_PSCR, KC_SCR, KC_PAUS,
```
Comment (fauxpark) [inline keyboards/mwstudio/mw80/keymaps/default/keymap.c:21]:
```suggestion
		KC_ESC, KC_F1, KC_F2, KC_F3, KC_F4, KC_F5, KC_F6, KC_F7, KC_F8,   KC_F9, KC_F10,  KC_F11,  KC_F12,  KC_F13,  KC_PSCR, KC_SCRL, KC_PAUS,
```
Sorry, typo.

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/mwstudio/mw80/info.json [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+        "console": false,
+        "extrakey": true,
+        "mousekey": true,
+        "nkro": true
+    },
+    "matrix_pins": {
+        "cols": ["B3", "A15", "A10", "A9", "A8", "B15", "B14", "A7", "B0", "B1", "B10", "A6", "A5", "A4", "A3", "A2", "A1" ],
+        "rows": ["B4", "B5", "B6", "B13", "B12", "B11" ]
+    },
+    "processor": "STM32F103",
```
Comment (fauxpark) [inline keyboards/mwstudio/mw80/info.json:18]:
```suggestion
    "processor": "STM32F103",
    "bootloader": "stm32-dfu",
```

File: keyboards/mwstudio/mw80/apm32f103/rules.mk [hunk 21]
```diff
@@ -0,0 +1,9 @@
+# MCU name
+MCU = STM32F103
+
+# Bootloader selection
+BOOTLOADER = stm32duino
+
+RGB_MATRIX_DRIVER = WS2812
```
Comment (fauxpark) [inline keyboards/mwstudio/mw80/apm32f103/rules.mk:7]:
```suggestion
```
Duplicated in info.json.

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by TW59420** [comment]
Because I think the atmosphere light can also use the matrix lighting effect, I think this way of writing is fine, the atmosphere light just doesn't correspond to the matrix

File: keyboards/mwstudio/mw80/info.json [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+    "bootloader": "stm32-dfu",
```
Comment (fauxpark) [inline keyboards/mwstudio/mw80/info.json:19]:
```suggestion
    "bootloader": "stm32duino",
```
Sorry again.

**Pull request closed without merging**